### PR TITLE
Don't send cached envelopes when rate-limiting is active

### PR DESCRIPTION
**Pull request by markushi**
## :scroll: Description
Expose RateLimiter via Hub to let SendCachedEnvelope integrations decide if they should send envelopes now or later.

## :bulb: Motivation and Context
Fixes https://github.com/getsentry/sentry-java/issues/1926

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry-android-core/src/main/java/io/sentry/android/core/SendCachedEnvelopeIntegration.java [hunk 1]
```diff
@@ -81,6 +83,15 @@ private synchronized void sendCachedEnvelopes(
       return;
     }
 
+    // in case there's rate limiting active, skip processing
+    final @Nullable RateLimiter rateLimiter = hub.getRateLimiter();
+    if (rateLimiter != null && rateLimiter.isActiveForCategory(DataCategory.All)) {
```
Comment (romtsn) [inline sentry-android-core/src/main/java/io/sentry/android/core/SendCachedEnvelopeIntegration.java:88]: `m`: I think we probably also have to add this to `EnvelopeFileObserver`, so if the hybrid SDKs are writing to the outbox folder we're also not trying to send envelopes if we're rate-limited right now

**Review by romtsn — APPROVED** [review]
Looking good, only one thing - maybe we should add the delay to avoid bursting our server with a bunch of events after a while and hitting another rate-limit? Similar to how it's done in https://github.com/getsentry/sentry-cocoa/pull/2263/files

File: sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java [hunk 2]
```diff
@@ -112,6 +113,17 @@ private synchronized void sendCachedEnvelopes(@NotNull IHub hub, @NotNull Sentry
       return;
     }
 
+    // in case there's rate limiting active, skip processing
```
Comment (adinauer) [inline sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java:116]: could we still run into problems, where we're not rate limited before starting to send a ton of events, then halfway through we get rate limited and still run into problems? Should this (also) be checked in or around e.g. `processFile`?

Comment (romtsn) [reply sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java:116]: yeah I think it's a good idea to check inside the for-loop 👍

File: sentry/src/main/java/io/sentry/IHub.java [hunk 3]
```diff
@@ -633,4 +634,7 @@ TransactionContext continueTrace(
    */
   @Nullable
   BaggageHeader getBaggage();
+
+  @Nullable
+  RateLimiter getRateLimiter();
```
Comment (stefanosiano) [inline sentry/src/main/java/io/sentry/IHub.java:639]: do we want to make it Internal? or is it good to provide it to the user?

Comment (romtsn) [reply sentry/src/main/java/io/sentry/IHub.java:639]: good point 👍 I think it doesn't make sense to expose it if it's a final class and not an interface, so I'll mark it as internal for now

**Review by stefanosiano — APPROVED** [review]

**Pull request merged**